subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/DependencyGraphBuilder.java

```diff
@@ -347,7 +350,13 @@ private void validateGraph(ResolveState resolveState) {
             ComponentState selected = module.getSelected();
             if (selected != null) {
                 if (selected.isRejected()) {
-                    GradleException error = new GradleException(new RejectedModuleMessageBuilder().buildFailureMessage(module));
+                    Pair<Capability, Collection<NodeState>> capabilityConflict = selected.getRejectedForCapabilityConflict();
```

Comment:
:o: Wouldn't it make sense to delegate this error message building to `NodeState` itself. The `if` here is a bit fishy, I feel like it should be handled on the "thing" that knows why it was rejected.